source/common/filter/filter_config_discovery_impl.h

```diff
@@ -0,0 +1,182 @@
+#include <unordered_set>
+
+#include "envoy/config/core/v3/extension.pb.h"
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/config/subscription.h"
+#include "envoy/filter/filter_config_provider.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/config/subscription_base.h"
+#include "common/init/manager_impl.h"
+#include "common/init/target_impl.h"
+
+namespace Envoy {
+namespace Filter {
+
+class HttpFilterConfigProviderManagerImpl;
+class HttpFilterConfigSubscription;
+
+using HttpFilterConfigSubscriptionSharedPtr = std::shared_ptr<HttpFilterConfigSubscription>;
+
+/**
+ * Implementation of a filter config provider using discovery subscriptions.
+ **/
+class DynamicFilterConfigProviderImpl : public HttpFilterConfigProvider {
+public:
+  DynamicFilterConfigProviderImpl(HttpFilterConfigSubscriptionSharedPtr&& subscription,
+                                  const std::set<std::string>& require_type_urls,
+                                  Server::Configuration::FactoryContext& factory_context);
+  ~DynamicFilterConfigProviderImpl() override;
+
+  // Config::ExtensionConfigProvider
+  const std::string& name() override;
+  absl::optional<Http::FilterFactoryCb> config() override;
+  void validateConfig(const ProtobufWkt::Any& proto_config,
+                      Server::Configuration::NamedHttpFilterConfigFactory&) override;
+  void onConfigUpdate(Http::FilterFactoryCb config, const std::string&) override;
+
+private:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig() : config_{absl::nullopt} {}
+    absl::optional<Http::FilterFactoryCb> config_{};
+  };
+
+  HttpFilterConfigSubscriptionSharedPtr subscription_;
+  const std::set<std::string> require_type_urls_;
+  ThreadLocal::SlotPtr tls_;
+
+  // Local initialization target to ensure that the subscription starts in
+  // case no warming is requested by any other filter config provider.
+  Init::TargetImpl init_target_;
+
+  friend class HttpFilterConfigProviderManagerImpl;
+};
+
+/**
+ * All filter config discovery stats. @see stats_macros.h
+ */
+#define ALL_FILTER_CONFIG_DISCOVERY_STATS(COUNTER)                                                 \
+  COUNTER(config_reload)                                                                           \
+  COUNTER(config_fail)
+
+/**
+ * Struct definition for all filter config discovery stats. @see stats_macros.h
+ */
+struct FilterConfigDiscoveryStats {
+  ALL_FILTER_CONFIG_DISCOVERY_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+/**
+ * A class that fetches the filter configuration dynamically using the filter config discovery API.
+ * Subscriptions are shared between the filter config providers. The filter config providers are
+ * notified when a new config is accepted.
+ */
+class HttpFilterConfigSubscription
```

Comment:
nit: can we split this out so the HTTP filter stuff is in the http directory with the core extension discovery code in a different directory? I'm not sure that filter is the right one now that it is generic?

Reply:
I don't have a good idea how much of the code is reusable in other extension contexts. I think much of it could be used for network filters but not sure about access log extension or others. Maybe I can just rename the folder to "extension"? If I move HTTP stuff somewhere else, there won't be much left here.

Reply:
So maybe just move it all into http for now and we can split later? If there is common code you could put in "extension"?

Reply:
does this look better?